src/qibo/transpiler/unroller.py

```diff
@@ -0,0 +1,460 @@
+        else:
+            if cz_dec.count_2q(gate) < iswap_dec.count_2q(gate):
+                return cz_dec(gate)
+            elif cz_dec.count_2q(gate) > iswap_dec.count_2q(gate):
+                return iswap_dec(gate)
+            # If equal check the decomposition with less 1 qubit gates.
+            # This is never used for now but may be useful for future generalization
+            elif cz_dec.count_1q(gate) < iswap_dec.count_1q(gate):  # pragma: no cover
+                return cz_dec(gate)
+            else:  # pragma: no cover
+                return iswap_dec(gate)
+    elif native_gates is NativeType.CZ:
+        return cz_dec(gate)
+    elif native_gates is NativeType.iSWAP:
+        if gate.__class__ in iswap_dec.decompositions:
+            return iswap_dec(gate)
+        else:
+            # First decompose into CZ
+            cz_decomposed = cz_dec(gate)
+            # Then CZ are decomposed into iSWAP
+            iswap_decomposed = []
+            for g in cz_decomposed:
+                # Need recursive function as gates.Unitary is not in iswap_dec
+                for g_translated in translate_gate(g, NativeType.iSWAP):
+                    iswap_decomposed.append(g_translated)
+            return iswap_decomposed
+    else:  # pragma: no cover
+        raise_error(NotImplementedError, "Use only CZ and/or iSWAP as native gates")
+
+
+class GateDecompositions:
+    """Abstract data structure that holds decompositions of gates."""
+
+    def __init__(self):
+        self.decompositions = {}
+
+    def add(self, gate, decomposition):
+        """Register a decomposition for a gate."""
+        self.decompositions[gate] = decomposition
+
+    def count_2q(self, gate):
+        """Count the number of two-qubit gates in the decomposition of the given gate."""
+        if gate.parameters:
+            decomposition = self.decompositions[gate.__class__](gate)
+        else:
+            decomposition = self.decompositions[gate.__class__]
+        return len(tuple(g for g in decomposition if len(g.qubits) > 1))
+
+    def count_1q(self, gate):
+        """Count the number of single qubit gates in the decomposition of the given gate."""
+        if gate.parameters:
+            decomposition = self.decompositions[gate.__class__](gate)
+        else:
+            decomposition = self.decompositions[gate.__class__]
+        return len(tuple(g for g in decomposition if len(g.qubits) == 1))
+
+    def __call__(self, gate):
+        """Decompose a gate."""
+        decomposition = self.decompositions[gate.__class__]
+        if callable(decomposition):
+            decomposition = decomposition(gate)
+        return [
+            g.on_qubits({i: q for i, q in enumerate(gate.qubits)})
+            for g in decomposition
+        ]
+
+
+onequbit_dec = GateDecompositions()
+onequbit_dec.add(gates.H, [gates.U3(0, 7 * np.pi / 2, np.pi, 0)])
+onequbit_dec.add(gates.X, [gates.U3(0, np.pi, 0, np.pi)])
```

Comment:
The $X$ gate should be done as one pulse like $R_{X}(\\pi)$. By decomposing it as a $U_{3}$, it is going to be done as two pulses like $R_{X}(\\pi/2)$. That introduces a lot of unnecessary noise.

Reply:
I already know this issue but when this unroller was first coded the compiler only accepted U3 and RZ as single qubit native gates. @stavros11 has this issue been solved?

Reply:
Indeed, right now the [compiler only has rules](https://github.com/qiboteam/qibolab/blob/main/src/qibolab/compilers/default.py) for `Z`, `RZ`, `GPI2` and `U3` so it will fail if we don't transpile the `X`. 
This is easy to fix, however I would prefer to have a more concrete solution, preferably in a different PR, where we decide the set of single-qubit native gates and we sync both the transpiler and compiler to respect that.